### ISSUE TITLE
Random out-of-range exceptions

### ISSUE THREAD

**Issue by FLamparski** [jasonrbriggs/stomp.py#33]
I am getting the following exception when running my stomp.py based client for a while. There are no other warning signs. I am trying to consume the UK's Network Rail datafeeds which use ActiveMQ as the backend.

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.4/threading.py", line 868, in run
    self._target(*self._args, **self._kwargs)
  File "/home/filip/RailDelay/server/lib/python3.4/site-packages/stomp/transport.py", line 299, in __receiver_loop
    f = utils.parse_frame(frame)
  File "/home/filip/RailDelay/server/lib/python3.4/site-packages/stomp/utils.py", line 131, in parse_frame
    f.cmd = preamble_lines[first_line]
IndexError: list index out of range

```

I am using the latest version of the package as published to pypi.

**Comment by jasonrbriggs** [comment]
Is it consistently failing (i.e. consistent after a time).  I've not seen it in testing, so difficult to know whether to just catch the error and log the problematic frame, or something else...

**Comment by FLamparski** [comment]
So after some testing it seems that the issue is very intermittent and possibly related to available client throughput. If I run into it again and can reliably reproduce, I'll let you know, but for now I'll close this ticket.

**Comment by aelnaggar** [comment]
can we maybe reopen this issue, I have it happening constantly : 

I added some prints, and seems that that frame has no info ,

frame info {cmd=None,headers=[{}],body=}

**Comment by jasonrbriggs** [comment]
What version of stomp.py, what's the message server (and version), and can you provide the actual stack trace?

**Comment by aelnaggar** [comment]
@jasonrbriggs  is there anyway to use openwire using this client api with instead of stomp ?

**Comment by jasonrbriggs** [comment]
Not without re-implementing the transport to support openwire rather than the stomp protocol.

**Comment by coreground** [comment]
I'm also seeing this error...

DEBUG:stomp.py:Sending frame ['ACK', '\n', 'message-id:broker-237112\n', 'subscription:/queue/pullconfig-1460394692\n', '\n', '\x00']
DEBUG:__main__:ack: message id broker-237112 with subscription /queue/pullconfig-1460394692
DEBUG:stomp.py:Received frame: 'MESSAGE', headers={'content-length': '131', 'expires': '1460481238075', 'destination': '/queue/pullconfig', 'user': '10.20.55.60,netbot,workers', 'ttl': '86400000', 'message-id': 'broker-237112', 'subscription': '/queue/pullc-1460394692'}, body='{"node": "eslonb", "type": "esl", "config": "/data/eslonb.cfg", "action": "etp", "timestamp": 1460394840}'
INFO:stomp.py:Receiver loop ended
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Library/Python/2.7/site-packages/stomp/transport.py", line 309, in __receiver_loop
    f = utils.parse_frame(frame)
  File "/Library/Python/2.7/site-packages/stomp/utils.py", line 125, in parse_frame
    f.cmd = preamble_lines[first_line]
IndexError: list index out of range

**Comment by paolodenti** [comment]
same error here, both on os x and linux, version 4.1.5

```
# python simple.py 
Exception in thread StompReceiverThread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/stomp/transport.py", line 319, in __receiver_loop
    f = utils.parse_frame(frame)
  File "/usr/local/lib/python2.7/dist-packages/stomp/utils.py", line 147, in parse_frame
    f.cmd = preamble_lines[first_line]
IndexError: list index out of range

```

simply.py

```

import time
import sys

import stomp

conn = stomp.Connection([('127.0.0.1', 61616)])
conn.start()
conn.connect('admin', 'admin', wait=True)

conn.send(body=' '.join(sys.argv[1:]), destination='/queue/test')

time.sleep(2)
conn.disconnect()

```

**Comment by jasonrbriggs** [comment]
I'm unable to reproduce -- even with the above example, it still doesn't fail (no matter how many times I tried). I'll add a check to that code to try and handle the problem in v4.1.16. Hopefully that will sort.

**Comment by paolodenti** [comment]
with 4.1.16 I cannot reproduce the same behaviour, this is what I get from logging

```

#: python simple.py
WARNING:stomp.py:Unknown response frame type: '�activemq' (frame length was 11)
WARNING:stomp.py:Unknown response frame type: '
                                               ' (frame length was 2)
WARNING:stomp.py:Unknown response frame type: 'o' (frame length was 2)
WARNING:stomp.py:Unknown response frame type: '
                                               ' (frame length was 1)
WARNING:stomp.py:Unknown response frame type: 'tcpnodelayenabled' (frame length was 20)
WARNING:stomp.py:Unknown response frame type: 'sizeprefixdisabled' (frame length was 20)
WARNING:stomp.py:Unknown response frame type: '	cachesize' (frame length was 11)
WARNING:stomp.py:Unknown response frame type: '' (frame length was 1)
WARNING:stomp.py:Unknown response frame type: '
                                               providername	' (frame length was 14)
WARNING:stomp.py:Unknown response frame type: activemq' (frame length was 9)
WARNING:stomp.py:Unknown response frame type: 'stacktraceenabled' (frame length was 20)
WARNING:stomp.py:Unknown response frame type: 'platformdetails	' (frame length was 17)
WARNING:stomp.py:Unknown response frame type: 'jjvm: 1.8.0_91, 25.91-b14, oracle corporation, os: linux, 4.9.4-moby, amd64' (frame length was 75)
WARNING:stomp.py:Unknown response frame type: '
                                               cacheenabled' (frame length was 15)
WARNING:stomp.py:Unknown response frame type: 'tightencodingenabled' (frame length was 23)
WARNING:stomp.py:Unknown response frame type: '
                                               maxframesize' (frame length was 14)
WARNING:stomp.py:Unknown response frame type: '@' (frame length was 2)
WARNING:stomp.py:Unknown response frame type: 'maxinactivityduration' (frame length was 23)
WARNING:stomp.py:Unknown response frame type: 'u0' (frame length was 2)
WARNING:stomp.py:Unknown response frame type: ' maxinactivitydurationinitaldelay' (frame length was 34)
WARNING:stomp.py:Unknown response frame type: ''' (frame length was 2)
WARNING:stomp.py:Unknown response frame type: 'providerversion	' (frame length was 17)

```

**Comment by bazell** [comment]
Was this issue ever resolved?  I have the same problem when running 4.1.17 on linux and trying to connect to a remote activemq broker.  I get a list of warnings very similar to the above.

**Comment by jasonrbriggs** [comment]
I was never able to reproduce it, and since I last tried, activemq doesn't seem to install any more on my test debian server -- apache apollo installs fine and I don't see the problem there at all.

**Comment by hs1211** [comment]
I also encounter this problem with amazon-mq which use activemq internally.

If anyone who resolved this problem. 

please share it one more time.

thanks in advance.

`
Unknown response frame type: '' (frame length was 3)
Traceback (most recent call last):
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 1591, in <module>
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 1018, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/hsboee/private/active-mq/test.py", line 18, in <module>
    conn.connect('id', 'password', wait=True)
  File "/Users/hsboee/private/active-mq/.venv/lib/python3.6/site-packages/stomp/protocol.py", line 126, in connect
    self.transport.wait_for_connection()
  File "/Users/hsboee/private/active-mq/.venv/lib/python3.6/site-packages/stomp/transport.py", line 318, in wait_for_connection
    raise exception.ConnectFailedException()
stomp.exception.ConnectFailedException
`

**Comment by amihalik** [comment]
hey @hs1211 I had the same issue.  Turns out I wasn't configuring SSL properly.  This works for me on AWS using Amazon MQ (Active MQ):

```
    c = stomp.Connection([(host, port)])
    c.set_listener('', stomp.PrintingListener())
    c.set_ssl(for_hosts=[(host, port)], ssl_version=ssl.PROTOCOL_TLS)
    c.start()

    c.connect(login, password, wait=True)
```

**Comment by grinkarthi** [comment]
Hi @amihalik 
I am using the snippet mentioned by you but I am not able to send message to the queue,
            client = stomp.Connection([(activemq_ip,
                                        int(activemq_port))])
            client.set_listener('', stomp.PrintingListener())
            client.set_ssl(for_hosts=[(activemq_ip, activemq_port)], ssl_version=ssl.PROTOCOL_TLSv1)
            client.start()
            print(client)
            client.connect(activemq_user_name, activemq_password, wait=True)
            client.send(activemq_queue_name, queue_message,
                        headers={'content-length': None},
                        persistent='false')
            client.disconnect()

I am using stomp version 4.1.11
Appreciate your support.

**Comment by amihalik** [comment]
I'm using the following code in an AWS Lambda (Python 3.6) to write messages from API Gateway to Amazon MQ.  The differences I notice are:

1. I'm using the latest stomp.py (4.1.21)
2. my port is a string and I use the full host name 
3. I'm using  ```ssl_version=ssl.PROTOCOL_TLS```
4. I don't pass in ```headers``` and ```persistent``` to ```client.send()```

I found it useful to set up Active MQ on my local machine to test and debug my stomp.py calls.  That's how I figured out the correct ```ssl_version```.  I can view the Active MQ logs on my local machine, but I cannot figure out how to view the Active MQ logs on Amazon MQ.  It appears that this is still an [outstanding feature request](https://forums.aws.amazon.com/thread.jspa?messageID=816372).

```
    c = stomp.Connection([(host, port)])
    c.set_listener('', stomp.PrintingListener())
    c.set_ssl(for_hosts=[(host, port)], ssl_version=ssl.PROTOCOL_TLS)
    c.start()

    c.connect(login, password, wait=True)
    c.send(body=message, destination=queue)
    c.disconnect()
```
